File: cpp/randomizedSet.cpp
#include <unordered_set>
#include <random>
#include <iterator>

using namespace std;

class RandomizedSet {
    unordered_set<int> s;
public:
    RandomizedSet() {
        
    }
    
    bool insert(int val) {
        if (s.find(val) != s.end()) {return false;}
        s.insert(val);
        return true;
    }
    
    bool remove(int val) {
        if (s.find(val) != s.end()) {
            s.erase(s.find(val));
            return true;
        }
        return false;
    }
    
    int getRandom() {
        return *next(s.begin(), rand()%s.size());
    }
};

/**
 * Your RandomizedSet object will be instantiated and called as such:
 * RandomizedSet* obj = new RandomizedSet();
 * bool param_1 = obj->insert(val);
 * bool param_2 = obj->remove(val);
 * int param_3 = obj->getRandom();
 */

File: rs/randomizedSet.rs
use std::collections::HashMap;
use rand::seq::SliceRandom;

struct RandomizedSet {
    hash: HashMap<i32, usize>,
    v:Vec<i32>
}

impl RandomizedSet {

    fn new() -> Self {
        Self {
            hash: HashMap::new(),
            v: Vec::new()
        }
    }
    
    fn insert(&mut self, val: i32) -> bool {
        if self.hash.contains_key(&val) {return false};
        self.hash.insert(val, self.v.len());
        self.v.push(val);
        true
    }
    
    fn remove(&mut self, val: i32) -> bool {
        match self.hash.remove(&val) {
            None => false,
            Some(i) => {
                self.v.swap_remove(i);
                if i < self.v.len() {
                    self.hash.insert(self.v[i], i);
                }
                true
            }
        }
    }
    
    fn get_random(&self) -> i32 {
        *self.v.choose(&mut rand::thread_rng()).unwrap()
    }
}

/**
 * Your RandomizedSet object will be instantiated and called as such:
 * let obj = RandomizedSet::new();
 * let ret_1: bool = obj.insert(val);
 * let ret_2: bool = obj.remove(val);
 * let ret_3: i32 = obj.get_random();
 */